lucene/lucene-directory/src/main/java/org/infinispan/lucene/cacheloader/LuceneCacheLoader.java

```diff
@@ -91,38 +92,20 @@ public boolean contains(final Object key) {
    }
 
    @Override
-   public void process(final KeyFilter filter, final CacheLoaderTask task, Executor executor, boolean fetchValue, boolean fetchMetadata) {
-      scanForUnknownDirectories();
-      ExecutorAllCompletionService eacs = new ExecutorAllCompletionService(executor);
-
-      final TaskContextImpl taskContext = new TaskContextImpl();
-      for (final DirectoryLoaderAdaptor dir : openDirectories.values()) {
-         eacs.submit(new Callable<Void>() {
-            @Override
-            public Void call() throws Exception {
-               try {
-                  final HashSet<MarshalledEntry> allInternalEntries = new HashSet<>();
-                  dir.loadAllEntries(allInternalEntries, Integer.MAX_VALUE, ctx.getMarshaller());
-                  for (MarshalledEntry me : allInternalEntries) {
-                     if (taskContext.isStopped())
-                        break;
-                     if (filter == null || filter.accept(me.getKey())) {
-                        task.processEntry(me, taskContext);
-                     }
-                  }
-                  return null;
-               }
-               catch (Exception e) {
-                  log.errorExecutingParallelStoreTask(e);
-                  throw e;
-               }
-            }
-         });
-      }
-      eacs.waitUntilAllCompleted();
-      if (eacs.isExceptionThrown()) {
-         throw new PersistenceException("Execution exception!", eacs.getFirstException());
-      }
+   public Publisher<MarshalledEntry<K, V>> publishEntries(Predicate<? super K> filter, boolean fetchValue, boolean fetchMetadata) {
+      return Flowable.fromIterable(openDirectories.values())
+            // Make sure that we update directories before a publisher is used
+            .doOnSubscribe(s -> scanForUnknownDirectories())
+            // We parallelize this since the loading below is blocking
+            .parallel()
+            .runOn(Schedulers.from(ctx.getExecutor()))
+            .flatMap(dir -> {
+               final Set<MarshalledEntry<K, V>> allInternalEntries = new HashSet<>();
+               dir.loadAllEntries(allInternalEntries, Integer.MAX_VALUE, ctx.getMarshaller());
+               return Flowable.fromIterable(allInternalEntries);
+            })
+            .filter(me -> filter == null || filter.test(me.getKey()))
+            .sequential();
```

Comment:
The user can still do whatever they want to the resulting publisher here. The problem is with reactive streams is you can't call `onNext` from different threads concurrently. This code here is ensuring that `onNext` observes this contract - note it will most likely be called from different threads, just not at the same time.

The user can still observe these entries on any number of threads that they want. Note a publisher is not "used" until someone subscribes to it.